src/tests/suite/daos_obj.c

```diff
@@ -1458,24 +1459,49 @@ punch_simple_internal(void **state, daos_obj_id_t oid)
 			    DAOS_TX_NONE, &req);
 
 
-	/* TODO Record enumeration still under development, use lookup. */
-	print_message("Lookup non-punched records:\n");
+	/**
+	 * Lookup and enumerate records for both punched and non-punched
+	 * entries.
+	 */
+	print_message("Lookup & enumerate non-punched records:\n");
 	for (i = 0; i < PUNCH_NUM_KEYS; i++) {
 		if (i % 2 == 0)
 			num_rec_exts = PUNCH_SCM_NUM_EXTS;
 		else
 			num_rec_exts = PUNCH_NVME_NUM_EXTS;
 		size = PUNCH_IOD_SIZE * num_rec_exts;
 		D_ALLOC(rec_fetch, size);
-		print_message("\tlookup:dkey:%s, akey:'akey'\n", dkeys[i]);
+
+		/* Lookup all non-punched records */
 		lookup_single_with_rxnr(dkeys[i], "akey", /*idx*/0, rec_fetch,
 					PUNCH_IOD_SIZE, size, DAOS_TX_NONE,
 					&req);
 		assert_memory_equal(rec_fetch, data_buf, size);
+		/* Enumerate all non-punched records */
+		memset(&anchor_out, 0, sizeof(anchor_out));
+		total_keys = 0;
+		while (!daos_anchor_is_eof(&anchor_out)) {
+			daos_epoch_range_t	eprs[PUNCH_ENUM_NUM];
+			daos_recx_t		recxs[PUNCH_ENUM_NUM];
+
+			enum_num = PUNCH_ENUM_NUM;
+			enumerate_rec(DAOS_TX_NONE, dkeys[i], "akey", &size,
+				      &enum_num, recxs, eprs, &anchor_out, true,
+				      &req);
+
+			total_keys += enum_num;
+
+			if (enum_num == 0)
+				break;
+		}
```

Comment:
hmm, I thought you want to check total_keys by assert_int_equal, otherwise no point to counting total_keys? or I miss sth?

Reply:
I can add a check for this, currently the value was just being printed